### ISSUE TITLE
Investigate "consumption-driven" trigger for executor shutdown in dynamic allocation

### ISSUE THREAD

**Issue by erikerlandson** [apache-spark-on-k8s/spark#261]
The basic idea is that currently, the only trigger for executor shutdown is "idle time" - if an executor has been idle past the configured duration, dynamic allocation logic will shut it down.  This allows for the possibility that an executor can get shut down before all of its shuffle data has been consumed, which makes external shuffle service necessary to guarantee that executor thrashing will not kill a job.

I'm wondering if a different trigger for executor shutdown might side-step the need for external shuffle service, which in turn could vastly simplify the deployment of dynamic allocation in a cluster manager (such as kubernetes).

So the key problem to avoid is an executor being shut down before its shuffle data has been consumed.  Is it possible to make executors, running an "internal" shuffle service, aware of when their shuffle data has been consumed, and it is now safe to shut them down?

I picture the basic state flow as something like:
1. Dynamic allocation notices executor is idle (this is existing logic)
1. It places this executor in "non-accepting" state: the executor does not get new work to do
    1. I think there is something like this state already, but if not it could be added
1. Once the executor determines that all of its shuffle data is consumed, it can be shut down

A first question is to figure out whether knowledge about which shuffle data blocks have been read, and when all of them have been read, is feasible to add.

**Comment by lins05** [comment]
> Once the executor determines that all of its shuffle data is consumed, it can be shut down

But the shuffle data may be needed again in the future. The purpose of external shuffle service is to ensure the shuffle blocks won't be re-computed even if the executor has been killed due to inactivity.

**Comment by kimoonkim** [comment]
>> Once the executor determines that all of its shuffle data is consumed, it can be shut down

> But the shuffle data may be needed again in the future.

@lins05 Are you saying something like below?

- If a future RDD needs to be recovered due to lost executors, we may have to recompute it using shuffle data from an earlier RDD as input.

**Comment by erikerlandson** [comment]
@lins05 my theory is that **if** you can keep track of which shuffle blocks have been read (and therefore are no longer needed) you could detect when all its data was consumed, and so know when the executor would be safe to shut down.

The current system of setting an idle-time threshold can't guarantee that, because it isn't looking at the right thing (although by configuring a sufficiently long idle-time you can make the probability of data loss arbitrarily small).

**Comment by ssuchter** [comment]
Hm, there’s a couple interesting edge cases here…

1. All the blocks get shuffled, executor gets shut down. Then one of the
consumers fails and gets restarted. It needs to read again, causing a
restart of the upstream executor, and we have a cascading increase of
resources. I think we could live without this being solved.

2. There are two downstream consumers of a given stage’s data. The data
gets shuffled once to the first consumer, executor gets shut down, then the
data needs to get shuffled to the second consumer. I don’t think we can
live without this being solved. Therefore we’d have to not just count if
the blocks have been shuffled once, but rather either the right number of
times or to the right number of consumers.

Sean

**Comment by kimoonkim** [comment]
@erikerlandson @lins05 

> my theory is that if you can keep track of which shuffle blocks have been read (and therefore are no longer needed)

I think this will be true only if the output RDDs that consumed those shuffle blocks are checkpointed to HDFS or something. Otherwise, when the output RDDs are lost at a later time point, Spark would try to recompute them. The shuffle blocks would be in need again. If shuffle blocks are missing, Spark would try to recreate them by doing earlier transformations that produced those shuffle blocks. This can be quite expensive.

**Comment by erikerlandson** [comment]
@kimoonkim that is correct.  I agree that checkpointing would be a useful precaution, and arguably just another variation on preserving very expensive results, similar to checkpointing results on iterative computations which otherwise generate intractably deep lineages.

**Comment by lins05** [comment]
I agree with @kimoonkim . For example,

```scala
    val rdd = sc.textFile("/opt/spark/README.md")
    val rdd2 = rdd.flatMap(line => line.split(" ")).map(w => (w, 1)).reduceByKey(_+_)
    val v1 = rdd2.collect()
    val v2 = rdd2.collect()
```
The first call of rdd2.collect would run a map stage and a reduce stage. The second call would only run the reduce stage because spark is smart enough to skip the map stage since the shuffle blocks are already there.

**Comment by erikerlandson** [comment]
Interesting, so another way of looking at shuffle service is that it's also acting as a kind of _de facto_ checkpointing mechanism.

**Comment by kimoonkim** [comment]
Yes, and it's cheaper than checkpointing to HDFS. Note Spark is faster than map reduce in part because it does not write as often to HDFS as map reduce. Checkpointing more to HDFS would make Spark slower.

**Comment by erikerlandson** [comment]
I think that is a good point.  OTOH, I might flip this around and say "spark is faster because it is running the equivalent of a local checkpointing file-system, in the form of a shuffle service, for each of its executors.   We pay for that by having to stand up this ~local filesystem~ shuffle service on each node where an executor might want to run; which we've discovered makes deploying dynamic allocation a lot less frictionless than it might be in theory.

I'm not claiming one is inherently better than another, but I think it does imply that choosing between these two modes of operation is a reasonable tradeoff choice, and it might not be a huge loss, if you consider that either way we are engaging the services of a checkpointing filesystem, it's just that we don't habitually think of the shuffle service that way.  From that point of view, it's revealing that most of the technical issues around shuffle service have had to do with what amount to file system mounting considerations in kube.

**Comment by kimoonkim** [comment]
I like your line of thinking. It certainly allows us to consider the tradeoff in relative terms.
